### ISSUE TITLE
Switching cloud account (Production/Development) may move existing surveys to other storage

### ISSUE THREAD

**Issue by ghachey** [ghachey/pacific-emis-education-survey-tool#71]
This is a little tricky. Changing from Production to Development or Development to Production with existing files on the tablet. Ideal the files would be "migrated" across. However, the one tricky bit is that this is a new Google service account and all UUID would be re-created and thus would need to be changed (all photos in a XML survey would have to be updated to the new photos' UUID. But I want to still assess the feasibility and repercussions of this feature. It would be useful and sometime IT staff can forget to set to Production for actual production work.

**Comment by maratomega** [comment]
Hi @ghachey 
1) We have config in Firebase that switches default accounts Prod/Dev - it won't work? Meaning, in Firebase we can set up the default account to be Production instead of the current Dev.
2) Correct me if I'm wrong please, we need to migrate all surveys from Dev to Prod, if IT staff switches account from Dev to Prod? Should we delete the surveys from Dev after migration?
3) Also, please clarify the reason you want to migrate surveys from Production to Development. Should we delete the surveys from Production after migration?

**Comment by ghachey** [comment]
> * We have config in Firebase that switches default accounts Prod/Dev - it won't work? Meaning, in Firebase we can set up the default account to be Production instead of the current Dev.

No, leave the default as is for now.

> * Correct me if I'm wrong please, we need to migrate all surveys from Dev to Prod, if IT staff switches account from Dev to Prod? Should we delete the surveys from Dev after migration?

Yes, if a tablet is switched from Dev to Prod then all the "tablet's" surveys would be migrated from Dev to Prod. Good question though. When you say "delete the surveys from Dev after migration" I assume you mean only the surveys from the tablet that was switched from Dev to Prod. Then yes, I think this would be the desired behavior provided it is only done upon *successful* migration of surveys and their photos, etc.

> * Also, please clarify the reason you want to migrate surveys from Production to Development. Should we delete the surveys from Production after migration?

It is not a essential feature, but if the feature works one way I would assume it also works the other way. For example, someone uses a tablet in Production which was supposed to be used in a training (and thus in Development mode).

In either case, I think I will put this in lower priority for now to be discussed further. And if it does go forward under no circumstance would we do any testing or development using the Production <-> Development. We would need to create a third mode for example Training and we could do all the feature testing between Development <-> Training service accounts.

**Comment by ghachey** [comment]
Additional notes based on our meeting:

- The cloudfiles on the source google service account are not to be deleted after a successful migration. Meaning if a user has to switch a tablet from Development mode to Production mode. All XML surveys (and photos) are transferred to the Production account. The files on the Development service account can remain there untouched and will be cleaned up manually afterwards from the backend.
- The most common use case for this is a user forgets to put the tablet in Production mode before going to the field. All the field work is done in Development and user only realizes after when they return to head quarters. There they switch from Development to Production. But the logic should work in any direction (meaning Development to Production, Development to Training, Training to Development, etc.)
- Do all the testing from Development to Training (and you can also test Training to Development). DO NOT touch Production stuff as usual as the team is working in production.
- Remember, Google will create new UUIDs for all the migrated surveys and photos. This means the migrated XML surveys will need to be processed carefully to edit all their photo IDs in them.
- Pay attention to the cloud status icons in the app. The icon that shows red for not started sync, orange for syncing in progress and green for a fully terminated sync. Of course, when switching to another service account this syncing statuses should restart and be based on the migration syncing process. Makes sense?
- I have shared privately the new needed service account "keys" for the new Training mode.

A final IMPORTANT note. Do not touch anything in the configuration or anywhere this may affect current production usage. Pay EXTRA care on this.

**Comment by maratomega** [comment]
@ghachey Can you please create folders for the Training mode

**Comment by ghachey** [comment]
@maratomega What do you mean by create folders for the Training mode? Do you mean create the service accounts? If so, they are already created and I believe the keys were shared with you. If you mean the actual folders inside (the context and photos directory) I think you can just create those yourself. So for example, for the default context of Pacific EMIS there is a specific Training (and Production and Testing) service account specifically for this with a key to access it. You would create the folders:

- PACEMIS
- PACEMIS\photos

**Comment by maratomega** [comment]
@ghachey  it's possible to create the folders on google disc if we will bring back the functionality of creating the folders from the app (if app can not find folders on the disc it was creating the new one to upload the surveys and photos) but since you mentioned that the above functionality not needed and actually the wrong one, I suggest that you will create the folders to upload the surveys and photos when we are on the Training mode. 

We do not have access to the disc where we upload all the surveys

**Comment by ghachey** [comment]
@maratomega Any progress on this?

**Comment by maratomega** [comment]
Completed

**Comment by ghachey** [comment]
@maratomega I tried this. I tried switching from the default of Testing mode to the Training mode with the FEDEMIS build. It does seem to bring the survey and it photos to the new Training service account but any subsequent edit I do not see reflected. And that event after the file cloud icons is green and therefore sync completed. Have editing surveys that were moved from one cloud service account to another been tested?

**Comment by maratomega** [comment]
> @maratomega I tried this. I tried switching from the default of Testing mode to the Training mode with the FEDEMIS build. It does seem to bring the survey and it photos to the new Training service account but any subsequent edit I do not see reflected. And that event after the file cloud icons is green and therefore sync completed. Have editing surveys that were moved from one cloud service account to another been tested?

@ghachey did you have a chance to create folders for Training mode on the disc?

**Comment by ghachey** [comment]
> > @maratomega I tried this. I tried switching from the default of Testing mode to the Training mode with the FEDEMIS build. It does seem to bring the survey and it photos to the new Training service account but any subsequent edit I do not see reflected. And that event after the file cloud icons is green and therefore sync completed. Have editing surveys that were moved from one cloud service account to another been tested?
> 
> @ghachey did you have a chance to create folders for Training mode on the disc?

@maratomega Yes, I also comment about this at https://github.com/PacificEMIS/pacific-emis-education-survey-tool/issues/72#issuecomment-2099568038

**Comment by ghachey** [comment]
> [@maratomega](https://github.com/maratomega) I tried this. I tried switching from the default of Testing mode to the Training mode with the FEDEMIS build. It does seem to bring the survey and it photos to the new Training service account but any subsequent edit I do not see reflected. And that event after the file cloud icons is green and therefore sync completed. Have editing surveys that were moved from one cloud service account to another been tested?

This feature is complete and the raised concern in above comment has been verified to work correctly.
